sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcUtilTest.java

```diff
@@ -41,4 +53,126 @@ public void testGetPreparedStatementSetCaller() throws Exception {
+        JdbcUtil.JdbcReadWithPartitionsHelper.getPartitionsHelper(
+            TypeDescriptor.of(DateTime.class));
+    DateTime onlyPoint = DateTime.now();
+    List<KV<DateTime, DateTime>> expectedRanges =
+        Lists.newArrayList(KV.of(onlyPoint, onlyPoint.plusMillis(1)));
+    List<KV<DateTime, DateTime>> ranges =
+        Lists.newArrayList(helper.calculateRanges(onlyPoint, onlyPoint, 10L));
+    // It is not possible to generate any more than one range, because the lower and upper range are
+    // exactly the same.
+    // The range goes from the current DateTime to ONE MILISECOND AFTER.
+    // Because the query's filter statement is : WHERE column >= lowerBound AND column < upperBound.
+    assertEquals(1, ranges.size());
+    assertArrayEquals(expectedRanges.toArray(), ranges.toArray());
+  }
+
+  @Test
+  public void testDatetimePartitioningWithMultiKey() {
+    JdbcReadWithPartitionsHelper<DateTime> helper =
+        JdbcUtil.JdbcReadWithPartitionsHelper.getPartitionsHelper(
+            TypeDescriptor.of(DateTime.class));
+    DateTime lastPoint = DateTime.now();
+    // At least 10ms in the past, or more.
+    DateTime firstPoint = lastPoint.minusMillis(10 + new Random().nextInt(Integer.MAX_VALUE));
+    List<KV<DateTime, DateTime>> ranges =
+        Lists.newArrayList(helper.calculateRanges(firstPoint, lastPoint, 10L));
+    // DateTime ranges are able to work out 10-11 ranges because they split in miliseconds which is
+    // very small granularity.
+    assertThat(Double.valueOf(ranges.size()), closeTo(10, 1));
+  }
+
+  @Test
+  public void testLongPartitioningWithSingleKey() {
+    JdbcReadWithPartitionsHelper<Long> helper =
+        JdbcUtil.JdbcReadWithPartitionsHelper.getPartitionsHelper(TypeDescriptors.longs());
+    List<KV<Long, Long>> expectedRanges = Lists.newArrayList(KV.of(12L, 13L));
+    List<KV<Long, Long>> ranges = Lists.newArrayList(helper.calculateRanges(12L, 12L, 10L));
+    // It is not possible to generate any more than one range, because the lower and upper range are
+    // exactly the same.
+    // The range goes from the current Long element to ONE ELEMENT AFTER.
+    // Because the query's filter statement is : WHERE column >= lowerBound AND column < upperBound.
+    assertEquals(1, ranges.size());
+    assertArrayEquals(expectedRanges.toArray(), ranges.toArray());
+  }
+
+  @Test
+  public void testLongPartitioningNotEnoughRanges() {
+    JdbcReadWithPartitionsHelper<Long> helper =
+        JdbcUtil.JdbcReadWithPartitionsHelper.getPartitionsHelper(TypeDescriptors.longs());
+    // The minimum stride is one, which is what causes this sort of partitioning.
+    List<KV<Long, Long>> expectedRanges =
+        Lists.newArrayList(KV.of(12L, 14L), KV.of(14L, 16L), KV.of(16L, 18L), KV.of(18L, 21L));
+    List<KV<Long, Long>> ranges = Lists.newArrayList(helper.calculateRanges(12L, 20L, 10L));
+    // It is not possible to generate any more than one range, because the lower and upper range are
```

Comment:
pls remove, comment is wrong for this case

Reply:
Done